### PR TITLE
fix: Use immutable names for volumes in example templates

### PR DESCRIPTION
**Pull request by mafredri**
This contributes towards #3000, #3386

Related #3409

**NOTE:** When this is merged, our release notes should make a mention / warning about updating to our example templates as doing so will very likely result in data loss (due to volume name changes).

### CODE REVIEW

**Review by kylecarbs — APPROVED** [review]

**Review by deansheather — CHANGES_REQUESTED** [review]
Have these been tested? I expected some of these names have length limits which will be hit by stringing together two UUIDs. Why not only use the workspace ID instead of both the user ID and the workspace ID instead?

Also, I don't think these should be merged until we've decided what upgrade strategy we want to put into our changelog. Are we going to tell people that they need to make a new template and abandon their old workspaces?

**Comment by mafredri** [comment]
Dogfood thing fails due to tagging with branch name (containing `/`), tried renaming the branch but that closed this PR instead. 😔

@deansheather I've tested Docker, but have not tested others (I can do a test on DO, but don't have setups for any others). Hmm, we could probably stick to only using the workspace UUID though 👍🏻. One motivation behind using both is additional metadata, but that would be amended by e.g. #3048.

**Comment by mafredri** [comment]
> Have these been tested? I expected some of these names have length limits which will be hit by stringing together two UUIDs. Why not only use the workspace ID instead of both the user ID and the workspace ID instead?

I've made the change to single UUID. 👍🏻

> Also, I don't think these should be merged until we've decided what upgrade strategy we want to put into our changelog. Are we going to tell people that they need to make a new template and abandon their old workspaces?

They can keep using their existing templates if they wish, but they should be vary of updating volume names to match that of these updated examples because that will result in loss of volumes.

> **BREAKING CHANGE:** We have updated all example templates to use immutable UUID-based volume names (#3663). Note that changing volume names in an existing template will result in data loss for all workspaces that are updated. If you would like to update your templates to match, please delete all existing workspaces as a precaution to avoid sudden or unexpected data loss.
>
> See #3386 for more discussion around this topic.

Thoughts @deansheather?

**Comment by mafredri** [comment]
@deansheather ping on this https://github.com/coder/coder/pull/3663#issuecomment-1228497578

@deansheather @kylecarbs thoughts? Do we still want to merge this change? Or do we want to combine it with some other changes? Or do something else entirely?

**Comment by deansheather** [comment]
I'd change the text to say that we don't recommend updating templates, and instead renaming the old one and creating a new template so users can migrate their data across. Hopefully we don't need to do this again.

**Comment by mafredri** [comment]
Updated text:

> **BREAKING CHANGE:** We have updated all example templates to use immutable UUID-based volume names (#3663). Note that changing volume names in an existing template will result in data loss for all workspaces that are updated. We do not recommend updating existing templates to match because of the risk of losing user workspace volumes, instead, rename the old template and upload a new one that users can migrate to.
>
> See #3386 for more discussion around this topic.

File: examples/templates/docker/main.tf [hunk 1]
```diff
@@ -80,11 +80,7 @@ resource "docker_container" "workspace" {
   dns      = ["1.1.1.1"]
   # Use the docker gateway if the access URL is 127.0.0.1
   command = [
-    "sh", "-c",
-    <<EOT
-    trap '[ $? -ne 0 ] && echo === Agent script exited with non-zero code. Sleeping infinitely to preserve logs... && sleep infinity' EXIT
-    ${replace(coder_agent.main.init_script, "/localhost|127\\.0\\.0\\.1/", "host.docker.internal")}
-    EOT
+    "sh", "-c", replace(coder_agent.main.init_script, "/localhost|127\\.0\\.0\\.1/", "host.docker.internal")
```
Comment (deansheather) [inline examples/templates/docker/main.tf:83]: This may make debugging harder since the container will get immediately deleted AFAIK

**Review by deansheather — APPROVED** [review]

File: examples/templates/docker/main.tf (hunk 1 above)
Comment (mafredri) [reply examples/templates/docker/main.tf:83]: We already have a trap in the bootstrap script, I think this is a remnant from a past implementation before that change.

This one also uses `sleep infinity` which isn’t compatible with all docker images.

**Comment by ammario** [comment]
We can use the `ignore_changes = all` lifecycle attribute to merge this in a non-breaking way.

**Comment by mafredri** [comment]
> We can use the `ignore_changes = all` lifecycle attribute to merge this in a non-breaking way.

It could definitely be a good addition, I'll try it out (I'm a bit worried about how Terraform will interpret this for existing resources that lack it, which might not help with the non-breakage).

**Comment by ammario** [comment]
> > We can use the `ignore_changes = all` lifecycle attribute to merge this in a non-breaking way.
> 
> It could definitely be a good addition, I'll try it out (I'm a bit worried about how Terraform will interpret this for existing resources that lack it, which might not help with the non-breakage).

I tried it out already in our dogfood image and it doesn't invalidate the resource. The `lifecycle` block is not included in the state file, it's a "meta block" so to speak.

**Pull request closed without merging**